### DISCUSSION TITLE
Combined Setup of Repository Browser and Repository Server including GitLab PyPI

### DISCUSSION THREAD

**Discussion by eddybl** [simple-repository/simple-repository#5] [Ideas]
# Motivation

<img width="819" alt="Setup" src="https://github.com/simple-repository/simple-repository/assets/43214674/2e71e64b-024b-4c15-bb07-05fadebb3c3a">

So the above picture describes the setup I would like to realise and currently exploring if I can do that.

For now I try to get the server and the web service running. For test purposes I do that in a Docker compose setup.

# Docker Setup

Compose file:

```Dockerfile
services:
  frontend:
    container_name: web-service
    build:
      dockerfile: Dockerfile-frontend
      context: ./dockerfiles
    ports:
      - '8081:8081'
    
  backend:
    container_name: server 
    build:
      context: ./dockerfiles
      dockerfile: Dockerfile-backend
    ports: 
      - 5432:5432
 ```

Dockerfile web service:

```Dockerfile
FROM python:3.11-bookworm

RUN pip install simple-repository-browser

COPY ./templates /opt/simple-repository-browser/templates

# TODO: Warm up the cache.

ENV XDG_CACHE_DIR=/cache

# Make the cache directory. Note that in OpenShift, this is mounted anyway.
RUN mkdir -p $XDG_CACHE_DIR \
 && chmod a+wr $XDG_CACHE_DIR \
;

EXPOSE 8081

CMD simple-repository-browser \
    --port 8081 \
    --templates-dir /opt/simple-repository-browser/templates \
    <URL-TO-SIMPLE-REPOSITORY-SEVER> \
;
```

Dockerfile backend:

```Dockerfile
FROM python:3.11-bookworm

RUN pip install simple-repository-server==0.5.0

CMD simple-repository-server \
    --port 5432 \
    <PATH-TO-PUBLIC-GITLAB-PYPI-ENDPOINT>
```

(I use bookworm only for testing, as I am more familiar in a debian based system, final deployment would most likely be alpine linux based)

# Current Status

So in the first iteration I only want to test this with a public gitlab endpoint, but actually have a two simple questions:

- How do I tell the web service to use the server as the PyPI source? What is the address to use?
- How do I actually install something from the server, what is the adress for PIP to point to?

**Comment by pelson** [comment]
> How do I tell the web service to use the server as the PyPI source? What is the address to use?

This sounds like a docker compose networking question. https://docs.docker.com/compose/networking/ is a useful first source of info, but it seems to me that the URL will simply be `http://backend:5432/simple/` (within the frontend).

For the exposed service, you should be able to reach it outside of docker-compose with `http://localhost:5432/simple/`. Note you will need to inform ``pip`` that you trust it (`PIP_TRUSTED_HOST=localhost`).

There may be a problem with the host being set as discussed in https://github.com/simple-repository/simple-repository-server/issues/5. It will be fixed when the line at https://github.com/simple-repository/simple-repository-server/blob/be716cb814d2d828ed16f671cd795c88fdfa70dc/simple_repository_server/__main__.py#L36 is in a release.


Overall, your setup looks to be in fairly good shape :+1:
